src/frames.ts

```diff
@@ -566,12 +566,19 @@ export class Frame {
     return this._raceWithCSPError(async () => {
       if (url !== null)
         return (await context.evaluateHandleInternal(addScriptUrl, { url, type })).asElement()!;
+      let result;
       if (path !== null) {
         let contents = await util.promisify(fs.readFile)(path, 'utf8');
         contents += '//# sourceURL=' + path.replace(/\n/g, '');
-        return (await context.evaluateHandleInternal(addScriptContent, { content: contents, type })).asElement()!;
+        result = (await context.evaluateHandleInternal(addScriptContent, { content: contents, type })).asElement()!;
+      } else {
+        result = (await context.evaluateHandleInternal(addScriptContent, { content: content!, type })).asElement()!;
       }
-      return (await context.evaluateHandleInternal(addScriptContent, { content: content!, type })).asElement()!;
+      // Another round trip to the browser to ensure that we receive CSP error messages
+      // (if any) logged asynchronously in a separate task on the content main thread.
+      if (this._page._delegate.cspErrorsAsynchronousForInlineScipts)
+        await context.evaluateInternal(() => true);
```

Comment:
Have you found the source of this? Are we sure it is a single task?

Reply:
The console logging [happens](http://ffsearch.azurewebsites.net/#path=%2Fhome%2Fjoe%2Ffirefox%2Fdom%2Fsecurity%2FnsCSPContext.cpp&line=1512) in a _separate_ task on the main thread. My assumption here is that the second evaluate is handled in its own task after the logging, so there is still a room for false positives. I could add some hooks in the native part to report the error synchronously via juggler but I felt like the problem we are fixing doesn't warrant it.